### PR TITLE
Expose CSI_FSGROUP_POLICY as an operator option

### PR DESCRIPTION
**Pull request by fmount**
The operator is now able to configure the CSIDriver object applying a
FsGroupPolicy value defined as input. csidriver.yaml is not a static resource
anymore and the manila driver operator manifest is updated to support this new
option.  The manila controller introduces some logic to dynamically refresh the
CSIDriver resource and, this pattern can be extended for any other option we'd
like to support in the future. The CSI_FSGROUP_POLICY option is defined as an
environment variable for the operator, and it's forced to None if the operator
doesn't make an explicit override.

Signed-off-by: Francesco Pantano <fpantano@redhat.com>

### CODE REVIEW

**Comment by gouthampacha** [comment]
@EmilienM @mandre can you please take a look? This change allows the CSI driver to report the "FSGroupPolicy" as None by default - which is the behavior we need with most network-attached storage that manila supports.

File: pkg/controllers/manila/manila.go [hunk 1]
```diff
@@ -69,24 +72,61 @@ func NewManilaController(
...
+	csiConfigData["fsGroupPolicy"] = fsGroupPolicy
+
+	csiConfig := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.ManilaCsiConfig,
+			Namespace: util.CloudConfigNamespace,
+		},
+		Data: csiConfigData,
+	}
+	_, _, err := resourceapply.ApplyConfigMap(ctx, c.kubeClient.CoreV1(), c.eventRecorder, &csiConfig)
```
Comment (gnufied) [inline pkg/controllers/manila/manila.go:109]: most attributes of `CSIDriver` object are immutable, so if configmap changes the fsGroup policy - `CSIDriver` has to be deleted and recreated. If `None` is what we want as default - why not specify in the manifest template above, rather than syncing it from configmap?

Comment (fmount) [reply pkg/controllers/manila/manila.go:109]: Right, that field of the `CSIDriver` object is immutable (like the other spec fields), and if the `CSIDriver` object is deleted, as per this code it's recreated getting the `fsgroup` value defined in the `ConfigMap`.
This makes the `ConfigMap` an extra layer that is used to consume the config options we'd like to propagate to the driver, and establishes a pattern where the input are defined using the manifest (as done in the template above), but are collected in the `ConfigMap` that can be used for multiple purposes in a follow up work (e.g., extra logic in the `StorageClasses` sync).
However, if you think this layer is too much for now (e.g., more complexity for a little gain), we can rework the PR to remove the `ConfigMap` and sync it from the env.

Comment (fmount) [reply pkg/controllers/manila/manila.go:109]: @gnufied  I'm going to push an extra change to simplify the code and avoid having this extra `ConfigMap` layer.

Comment (fmount) [reply pkg/controllers/manila/manila.go]: I guess this part of the conversation can be now resolved. 

tl;dr: I removed the `ConfigMap`, and the `fsGroupPolicy` value is configured using the manifest env var section. Editing this field is something that can't be done as it's immutable (and not the goal of the PR); however, if the operator delete the `CSIDriver` resource and restart the operator with a different value, that object is recreated accordingly.

Comment (gnufied) [reply pkg/controllers/manila/manila.go]: There are several problems with approach proposed here:
1. You are using `ApplyCSIDriver` function - which will not allow you to change `fsGroupPolicy`, so once created `CSIDriver` object is immutable. 
2. 
3. Manila operator is a default Driver and I am not sure where will users edit the deployment template and specify `fsGroupPolicy` as environment variable. The changes to deployment template will be overridden by cluster-storage-operator.  This is not how customization of CVO managed operators work.

Comment (fmount) [reply pkg/controllers/manila/manila.go]: That's not the point: the goal is not giving the opportunity to the (human) operator to edit an immutable field, but to expose an env variable for corner cases where people need to set a value != None.
If you delete the resource `CSIDriver`, it gets recreated/reconciled with the value specified as input.
To make people able to change this behavior, they should stop the operator, delete the resource and restart it with a different value, but definitely not a path we're supporting by default.

File: pkg/controllers/manila/manila.go [hunk 2]
```diff
@@ -123,6 +145,65 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
...
+
+	crd := &storagev1.CSIDriver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "manila.csi.openstack.org",
+			Annotations: map[string]string{
+				"csi.openshift.io/managed": "true",
+			},
+		},
+		Spec: csiOpt,
+	}
```
Comment (jsafrane) [inline pkg/controllers/manila/manila.go:184]: Please read static yaml manifest from `assets/` and change only `fsGroupPolicy` here - I'd like to have all objects in one place (`assets/`) and not to look around various places.

File: manifests/07_deployment.yaml [hunk 3]
```diff
@@ -38,6 +38,8 @@ spec:
           valueFrom:
             fieldRef:
               fieldPath: metadata.name
+        - name: CSI_FSGROUP_POLICY
+          value: None
```
Comment (jsafrane) [inline manifests/07_deployment.yaml:42]: How exactly will users set this env. var? Right now the deployment is created by cluster-storage-operator and it does not have any option to configure that.

This PR opens a broader question how CSI drivers should be configured. Right now, we don't have any good API for that - `ClusterCSIDriver` does not have any driver-specific fields.

File: pkg/controllers/manila/manila.go (hunk 1 above)
Comment (jsafrane) [reply pkg/controllers/manila/manila.go]: > To make people able to change this behavior, they should stop the operator, delete the resource and restart it with a different value, but definitely not a path we're supporting by default.

If users disables cluster-storage-operator or set it to `Unmanaged`, so it does not overwrite Manila operator Deployment, then the cluster cannot be upgraded. This is good for local debugging, but not really usable in any production env.

File: manifests/07_deployment.yaml (hunk 3 above)
Comment (jsafrane) [reply manifests/07_deployment.yaml:42]: I see already discussion above, let's continue there (and please mark this as Resolved)

File: pkg/controllers/manila/manila.go (hunk 2 above)
Comment (fmount) [reply pkg/controllers/manila/manila.go:184]: That object is removed from assets/ because the spec field is computed in a function according to the env defined in the operator manifest. 
Restoring assets/ makes this PR pointless.
I can close it if this approach doesn't match the way the operator is managed.

Comment (jsafrane) [reply pkg/controllers/manila/manila.go:184]: Some parts of assets/controller.yaml (and other yaml files) are computed too. Still, the template is in assets and only the computed parts are modified in go code. I'd really like to have all the templates in assets and only set the env. var in go code.

Comment (fmount) [reply pkg/controllers/manila/manila.go:184]: Ok, thank you for the comment and the suggestion. 
After digging a bit more, I realized that you can still have the `csidriver.yaml` asset statically defined, and put some logic in library-go as done for `StorageClasses` [1]  to make sure the the whole object is recreated in case an "immutable" field changes.
`FsGroupPolicy` is one of them: we can build a generic static CSIDriver object, with the default value, and recreate it in case it changes.
I'll push an update to reflect this behavior, which is the same pattern I see for the `StorageClass`.

[1] https://github.com/openshift/library-go/commit/354d53d860bff27de6adbf72dfaa4812a9fff61d

File: pkg/controllers/manila/manila.go [hunk 4]
```diff
@@ -123,6 +145,74 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
...
+	klog.V(4).Infof("Starting CSI driver config refresh")
+	defer klog.V(4).Infof("CSI driver config refresh finished")
+
+	var errs []error
+
+	crd, err := c.kubeClient.StorageV1().CSIDrivers().Get(ctx, "manila.csi.openstack.org", metav1.GetOptions{})
+	crd.Spec = spec
+
+	if err != nil {
+		crd = c.generateCSIDriver(spec)
```
Comment (fmount) [inline pkg/controllers/manila/manila.go:171]: @jsafrane in L167 I get the existing CSIDriver object and patch it with the generated spec.
If an error occurs it means the object is not there, and I generate it with the existing function.

This shouldn't happen because as per your previous comment/review we want the operator to apply static resources when it starts, and `CSIDriver` object is one of them.

However, if I keep `csidriver.yaml` among the applied static resources [1], after updating the `CSISpec` map the operator reconciles it with the original object defined in the static file, which means we can't take advantage of [2] because after patching the object the `StaticResourceController` deletes and recreates it.

If I remove [1] and manually apply `csidriver.yaml` (or if I rely on the `generateCSIDriver` function), it works as expected.

Do you have a suggestion on how we can remove this `generateCSIDriver` function and apply the static resource defined in the `csidriver.yaml` file without making it part of the resources reconciled by the `StaticResourceController`?

[1] https://github.com/openshift/csi-driver-manila-operator/blob/master/pkg/operator/starter.go#L77
[2] https://github.com/openshift/library-go/pull/1324/commits/c4d0b28d0d214f590dd499961b1a55c209ef1822

File: pkg/controllers/manila/manila.go [hunk 5]
```diff
@@ -123,6 +145,75 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
...
+func (c *ManilaController) applyCSIDriver(ctx context.Context, spec storagev1.CSIDriverSpec) error {
+	klog.V(4).Infof("Starting CSI driver config refresh")
+	defer klog.V(4).Infof("CSI driver config refresh finished")
+
+	var errs []error
+
+	crd, err := c.kubeClient.StorageV1().CSIDrivers().Get(ctx, "manila.csi.openstack.org", metav1.GetOptions{})
+
+	if err != nil {
+		crd = c.generateCSIDriver()
```
Comment (fmount) [inline pkg/controllers/manila/manila.go:170]: @jsafrane in L167 I get the existing CSIDriver object and patch it with the generated spec.
If an error occurs it means the object is not there, and I generate it with the existing function.

This shouldn't happen because as per your previous comment/review we want the operator to apply static resources when it starts, and `CSIDriver` object is one of them.

However, if I keep `csidriver.yaml` among the applied static resources [1], after updating the `CSISpec` map the operator reconciles it with the original object defined in the static file, which means we can't take advantage of [2] because after patching the object the `StaticResourceController` deletes and recreates it.

If I remove [1] and manually apply `csidriver.yaml` (or if I rely on the `generateCSIDriver` function), it works as expected.

Do you have a suggestion on how we can remove this `generateCSIDriver` function and apply the static resource defined in the `csidriver.yaml` file without making it part of the resources reconciled by the `StaticResourceController`?

[1] https://github.com/openshift/csi-driver-manila-operator/blob/master/pkg/operator/starter.go#L77
[2] https://github.com/openshift/library-go/pull/1324/commits/c4d0b28d0d214f590dd499961b1a55c209ef1822

Comment (fmount) [reply pkg/controllers/manila/manila.go:170]: One way could be turning the `generateCSIDriver` function into something like:

```
func (c *ManilaController) generateCSIDriver() *storagev1.CSIDriver {

	sch := runtime.NewScheme()
	_ = scheme.AddToScheme(sch)
	_ = apiextv1beta1.AddToScheme(sch)
	decode := serializer.NewCodecFactory(sch).UniversalDeserializer().Decode
	stream, _ := ioutil.ReadFile("assets/csidriver.yaml")

	obj, _, _ := decode(stream, nil, nil)

	csi_obj := obj.(*storagev1.CSIDriver)

	return csi_obj
```
By doing this, we're getting the static asset and decoding it into a `storagev1.CSIDriver` object, so we keep the static asset definition, and patching it according to the `env` passed to the operator.
wdyt

Comment (jsafrane) [reply pkg/controllers/manila/manila.go:170]: There already is [ReadFile](https://github.com/openshift/csi-driver-manila-operator/blob/master/assets/assets.go#L11) that returns `[]byte` with a file from `assets/` and [ReadCSIDriverV1OrDie]( https://github.com/openshift/library-go/blob/3c66b317b110fe1a94bb0c9a9fa9f7a46df29941/pkg/operator/resource/resourceread/storage.go#L37) that parses `[]byte` it into a CSIDriver object.

And finally, there is [ApplyCSIDriver](https://github.com/openshift/library-go/blob/3c66b317b110fe1a94bb0c9a9fa9f7a46df29941/pkg/operator/resource/resourceapply/storage.go#L108), that creates or updates (re-creates) the CSI driver.

So this PR can be just a new function `syncCSIDriver` with few lines of code (+unit tests, that may be harder).

Comment (fmount) [reply pkg/controllers/manila/manila.go:170]: Nice thank you for the hint, I wasn't aware about these functions.
Let me simplify the PR further, as we can just have a syncCSIDriver() that relies on the underlying functions!

Thanks!

Comment (fmount) [reply pkg/controllers/manila/manila.go]: Hey @jsafrane, thanks for all your input so far, I managed to rely on the existing code as much as possible to sync the `CSIDriver` object and make the `FsGroupPolicy` a parameter that can be changed according to the env defined in the operator manifest.
As per your previous comment, this PR is now something that introduces the `syncCSIDriver` function that creates/updates the object according to the operator input. 

Here the main highlights:

* The CSIDriver resource is static and defined in `assets/`: the operator reads the file and build the object in case it doesn't exist, otherwise it applies it with the `CSIDriverSpec` passed as input;

* The `ApplyCSIDriver` function, provided in library-go, is used to update the immutable fields spec through deletion and recreation;

I guess we can now resolve this conversation, unless there's something I've overlooked.

File: pkg/controllers/manila/manila.go [hunk 6]
```diff
@@ -123,6 +147,66 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
...
+	if err != nil {
+		// no CSIDriverObject found, read from the asset and apply
+		stream, e := assets.ReadFile("csidriver.yaml")
+		if e != nil {
+			klog.V(4).Infof("Error loading the CSIDriver resource")
+		}
+		crd = resourceread.ReadCSIDriverV1OrDie(stream)
+	}
+
+	crd.Spec = spec
```
Comment (jsafrane) [inline pkg/controllers/manila/manila.go:180]: If `Get()` above returns error (e.g. CSIDriver does not exist), `crd` is nil at this point.

IMO what you want in the whole `syncCSIDriver()` is:

```go
stream, err := assets.ReadFile("csidriver.yaml")
if err != nil {
	panic("Error loading the CSIDriver resource", err)
}

cr = resourceread.ReadCSIDriverV1OrDie(stream)
cr.Spec.FSGroupPolicy = getFSGroupPolicy()

_, _, err = resourceapply.ApplyCSIDriver(ctx, c.kubeClient.StorageV1(), c.eventRecorder, cr)
return err
```

File: manifests/07_deployment.yaml (hunk 3 above)
Comment (jsafrane) [inline manifests/07_deployment.yaml:42]: How is an user supposed to set this env. var in a supported cluster? This operator Deployment will be created by cluster-storage-operator (CSO). If user changes env. var in the Deployment, CSO will overwrite it. They can stop CSO or set it `Unmanaged`, but then I would not call it a supported cluster (e.g. such a cluster cannot be upgraded).

File: pkg/controllers/manila/manila.go (hunk 6 above)
Comment (fmount) [reply pkg/controllers/manila/manila.go]: ack, so basically we're not reading the existing object anymore: we're just loading the asset and customizing the spec as defined in the manifest.
This is simple and less error prone: the last update defines this kind of behavior.

File: manifests/07_deployment.yaml (hunk 3 above)
Comment (fmount) [reply manifests/07_deployment.yaml:42]: Not sure how CSO works (just learning it right now), but I probably see your point: the human operator is not supposed to stop the `manila-csi-operator`, change the value and restart it again with a different env var value, while `cluster-storage-operator` is supposed to manage the lifecycle of the `manila-csi-operator`.
This means that this PR isn't enough to allow people to change the  `CSI_FSGROUP_POLICY` value, because it should be reflected in [1], that drives the behavior of this operator.
Now, the main goal of this patch is not changing the supported policy and allow people to move to a not supported configuration, and the default value (`None`) is enforced is nothing is specified; however, it would be great having this parameter exposed (it was immutable in the `CSIDriver` spec) and have the ability to change it under specific conditions.
So, let's say this patch is not "harmful" and introduces this logic of reading and assigning the `fsGroupPolicy` from an env var: adding the var in [1] is the right (next) move, or am I reading it wrong? 

[1] https://github.com/openshift/cluster-storage-operator/blob/master/assets/csidriveroperators/manila/07_deployment.yaml

Comment (jsafrane) [reply manifests/07_deployment.yaml:42]: [1] is indeed the canonical source of Manila operator Deployment. It is instantiated by CSO and here we have the same problem - how users tell CSO to add the env. var there?

If you want to have a supported API, then extension of ClusterCSIDriver would be probably appropriate.
If you want to have an unsupported hack for experiments, stopping CSO and users editing Manila CSI driver deployment manually could be OK.

File: pkg/controllers/manila/manila.go [hunk 7]
```diff
@@ -37,10 +40,11 @@ import (
...
 type ManilaController struct {
-	operatorClient     v1helpers.OperatorClient
-	kubeClient         kubernetes.Interface
-	openStackClient    *openStackClient
-	storageClassLister storagelisters.StorageClassLister
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	openStackClient      *openStackClient
+	storageClassLister   storagelisters.StorageClassLister
+	csiDriverClassLister storagelisters.CSIDriverLister
```
Comment (jsafrane) [inline pkg/controllers/manila/manila.go:47]: `csiDriverLister`

**Comment by jsafrane** [comment]
> @jsafrane - is it okay to get this in for test/dev purposes?

Ack
/approve
I let someone from @openshift/team-openstack-maintainers to lgtm

File: pkg/controllers/manila/manila.go [hunk 8]
```diff
@@ -123,6 +136,62 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
 	return c.setEnabledCondition(ctx)
 }
 
+func (c *ManilaController) getFsGroupPolicy(ctx context.Context) storagev1.FSGroupPolicy {
+
+	// Set the fsGroupPolicy param to the default value
+	fsGroupPolicy := "None"
```
Comment (mandre) [inline pkg/controllers/manila/manila.go:142]: The [doc for `fsGroupPolicy`]( https://kubernetes-csi.github.io/docs/support-fsgroup.html#csi-driver-fsgroup-support) says that:
> If undefined, fsGroupPolicy will default to ReadWriteOnceWithFSType, keeping the previous behavior.

`fsGroupPolicy` was previously undefined, so setting it to `None` here will change the behavior, isn't it?

File: pkg/controllers/manila/manila.go [hunk 9]
```diff
@@ -123,6 +136,62 @@ func (c *ManilaController) sync(ctx context.Context, syncCtx factory.SyncContext
...
+
+	// Try to get the current value defined for the operator
+	if c.isFSGroupPolicyAllowed(os.Getenv("CSI_FSGROUP_POLICY")) {
```
Comment (mandre) [inline pkg/controllers/manila/manila.go:145]: The logic is a bit confusing. IMO having a new function to validate the value makes it unnecessary complicate, we can do the validation here if we want to. We should also log for unknown values.

How about something like the following?
```go 
func (c *ManilaController) getFsGroupPolicy(ctx context.Context) storagev1.FSGroupPolicy {
	// NOTE: shoudn't the default be storagev1.ReadWriteOnceWithFSTypeFSGroupPolicy?
	fsGroupPolicy := storagev1.NoneFSGroupPolicy

	fsGroupPolicyFromEnv := storagev1.FSGroupPolicy(os.Getenv("CSI_FSGROUP_POLICY"))
	switch fsGroupPolicyFromEnv {
	case storagev1.NoneFSGroupPolicy, storagev1.FileFSGroupPolicy, storagev1.ReadWriteOnceWithFSTypeFSGroupPolicy:
		fsGroupPolicy = fsGroupPolicyFromEnv
	default:
		if fsGroupPolicyFromEnv != "" {
			klog.V(4).Infof("Invalid CSI_FSGROUP_POLICY %q. Ignoring.", fsGroupPolicyFromEnv)
		}
	}

	return fsGroupPolicy
}
```

File: pkg/controllers/manila/manila.go (hunk 8 above)
Comment (fmount) [reply pkg/controllers/manila/manila.go:142]: Exactly, but as per [1]  and the conversation in [2] the purpose if to set to `None` for manila by default.  

[1] https://github.com/kubernetes/cloud-provider-openstack/pull/1894
[2] https://github.com/kubernetes/cloud-provider-openstack/issues/1893

File: pkg/controllers/manila/manila.go (hunk 9 above)
Comment (fmount) [reply pkg/controllers/manila/manila.go:145]: +1 thanks, the logic you're proposing here looks good to me. Updating the patch in a bit with a more simplified logic.

File: pkg/controllers/manila/manila.go (hunk 7 above)
Comment (fmount) [reply pkg/controllers/manila/manila.go:47]: ack thanks

**Review by mandre — COMMENTED** [review]
Apologies for the delay in reviewing the patch, summer vacation happened.
/lgtm
